## WEEK 10	:	χ<sup>2</sup> Test and introduction to cluster analysis

### Chi-square of Independence

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving acad.xlsx to acad.xlsx


In [ ]:
acad = pd.read_excel('acad.xlsx')
print(acad)

    Rsp No  aa  pe  sm  ae  r  g  c
0        1  99  19   1   2  0  0  1
1        2  46  12   0   0  0  0  0
2        3  57  15   1   1  0  0  0
3        4  94  18   2   2  1  1  1
4        5  82  13   2   1  1  1  1
5        6  59  12   0   0  2  0  0
6        7  61  12   1   2  0  0  0
7        8  29   9   0   0  1  1  0
8        9  36  13   1   1  0  0  0
9       10  91  16   2   2  1  1  0
10      11  55  10   0   0  1  0  0
11      12  58  11   0   1  0  0  0
12      13  67  14   1   1  0  1  1
13      14  77  14   1   2  2  1  0
14      15  71  12   0   0  2  1  0
15      16  83  16   2   2  1  0  1
16      17  96  15   2   2  2  0  1
17      18  87  12   1   1  0  0  1
18      19  62  11   0   0  0  0  0
19      20  52   9   0   1  2  1  0
20      21  46  10   1   0  0  1  0
21      22  91  20   2   2  1  0  0
22      23  85  12   2   1  1  1  1
23      24  48  11   1   1  2  0  0
24      25  81  17   1   1  1  1  1
25      26  74  16   2   1  2  1  0
26      27  68  12   2   1  

Test the hypothesis that "gender and student motivation" are independent

In [ ]:
#cross table between gender and student's motivation
obs = pd.pivot_table(acad[['g', 'sm']], index='g', columns='sm', aggfunc=len)
obs

sm   0   1  2
g            
0   10  13  6
1    4   9  8

In [ ]:
# perform chi2 test to check the independence
from scipy.stats import chi2_contingency

In [ ]:
chi2, p, dof, tbl=chi2_contingency(obs)

In [ ]:
#table value
chi2

2.3649585225939904

In [ ]:
#calculated value
p

0.3065178579178871

In [ ]:
#degrees of freedom
dof

2

chi2>p hence accept null hypothesis that the gender and student motivation are independent

In [ ]:
#contingency table
tbl

array([[ 8.12, 12.76,  8.12],
       [ 5.88,  9.24,  5.88]])

### χ<sup>2</sup> Goodness of fit test

In [ ]:
import scipy
from scipy.stats import chi2
from scipy.stats import poisson

Checking GOF of Poisson distribution

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving P_distribution.xlsx to P_distribution.xlsx


In [ ]:
data = pd.read_excel("P_distribution.xlsx")
print(data)

    Arrivals  Frequency
0          0          0
1          1          1
2          2          4
3          3         10
4          4         14
5          5         20
6          6         12
7          7         12
8          8          9
9          9          8
10        10          6
11        11          3
12        12          1


In [ ]:
observed_freq = data['Frequency']
total_arrival = 600
total_time_period = 100
mu = total_arrival/total_time_period
expected_freq = []
mu

6.0

In [ ]:
for i in range(len(observed_freq)):
  E_freq = 100*poisson.pmf(i, mu)
  expected_freq.append(E_freq)

expected_freq

[0.24787521766663584,
 1.4872513059998145,
 4.461753917999444,
 8.923507835998894,
 13.385261753998332,
 16.062314104797995,
 16.06231410479801,
 13.767697804112569,
 10.32577335308442,
 6.883848902056284,
 4.130309341233764,
 2.2528960043093247,
 1.1264480021546681]

In [ ]:
expected_freq_round_off = [round(elem, 2) for elem in expected_freq]
expected_freq_round_off

[0.25,
 1.49,
 4.46,
 8.92,
 13.39,
 16.06,
 16.06,
 13.77,
 10.33,
 6.88,
 4.13,
 2.25,
 1.13]

In [ ]:
df = pd.DataFrame(list(zip(observed_freq, expected_freq_round_off)), columns=["Observed Frequency", "Expected Frequency"])
print(df)

    Observed Frequency  Expected Frequency
0                    0                0.25
1                    1                1.49
2                    4                4.46
3                   10                8.92
4                   14               13.39
5                   20               16.06
6                   12               16.06
7                   12               13.77
8                    9               10.33
9                    8                6.88
10                   6                4.13
11                   3                2.25
12                   1                1.13


Example 2: when expected freq already given

In [ ]:
obs_freq = [5, 10, 14, 20, 12, 12, 9, 8, 10]
exp_freq = [6.20, 8.92, 13.39, 16.06, 16.06, 13.77, 10.33, 6.88, 8.39]

Null hypothesis: Is there any significant difference between these parameters?

In [ ]:
scipy.stats.chisquare(obs_freq, exp_freq)

Power_divergenceResult(statistic=3.2738182931105193, pvalue=0.916017731732134)

hence the calculated value is 3.27

Checking GOF for uniform distribution

In [ ]:
from scipy.stats import chi2

In [ ]:
#when alpha is 5% chi2 table value is 14.6
chi2.ppf(0.95, 7)

14.067140449340169

Table value > calculated value hence accept null hypothesis.

In [ ]:
x=[1610,1585,1649,1590,1540,1397,1410,1350,1495,1564,1602,1655]
np.mean(x)

1537.25

In [ ]:
exp_f=[1537.25,1537.25,1537.25,1537.25,1537.25,1537.25,1537.25,1537.25,1537.25,1537.25,1537.25,1537.25]

In [ ]:
from scipy.stats import chisquare
chisquare(x,exp_f)

Power_divergenceResult(statistic=74.37583346885673, pvalue=1.78545252783034e-11)

In [ ]:
from scipy.stats import chi2
chi2.ppf(0.99,11)

24.724970311318277

Calculated value > table value hence reject null hypothesis. (74.37>24.725)

Checking GOF for normal distribution

In [ ]:
A=[33,43,44,45,52,52,56,58,63,64,64,65,66,68,70,72,73,73,74,75,83,84,85,86,91,92,94,98,102,105]
mean = np.mean(A)
print(mean)

71.0


In [ ]:
std = np.std(A)
print(std)

18.226354544998845


Null hypothesis: the population of number of units sold has a normal distribution with mean 71 and standard deviation of 18.23.

Alternate hypothesis: the population of number of units sold does not have a normal distribution with mean 71 and standard deviation of 18.23.

In [ ]:
x = 1/6 #for 6 equal probability intervals

In [ ]:
for j in range(1,6):
  prob_intervals = [scipy.stats.norm.ppf(j*x, mean, std)]
  print(prob_intervals)

[53.36743154175236]
[63.14941153083116]
[71.0]
[78.85058846916884]
[88.63256845824763]


In [ ]:
exp_freq = [5,5,5,5,5,5] #will divide the normal distribution into 6 intervals at frequency of 5 each
obs_freq = [6,3,6,5,4,6]
scipy.stats.chisquare(obs_freq, exp_freq)

Power_divergenceResult(statistic=1.5999999999999999, pvalue=0.9012493445012737)

### Clustering Analysis

### Assignment question: solved

In [ ]:
exp_freq = [200,200,200,200,200] #will divide the normal distribution into 6 intervals at frequency of 5 each
obs_freq = [180,250,120,225,225]
scipy.stats.chisquare(obs_freq, exp_freq)

Power_divergenceResult(statistic=52.75, pvalue=9.612518035368181e-11)

In [ ]:
from scipy.stats import chi2
chi2.ppf(0.95,4)

9.487729036781154